src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -0,0 +1,365 @@
+        return recomputeNext();
+    }
+
+    protected abstract void performSkipTo(PrimaryKey nextToken);
+
+    protected PrimaryKey recomputeNext()
+    {
+        return tryToComputeNext() ? peek() : endOfData();
+    }
+
+    protected boolean tryToComputeNext()
+    {
+        boolean hasNext = super.tryToComputeNext();
+        current = hasNext ? next : getMaximum();
+        return hasNext;
+    }
+
+    public static RangeIterator empty()
+    {
+        return EmptyRangeIterator.instance;
+    }
+
+    public static class EmptyRangeIterator extends RangeIterator
+    {
+        static final RangeIterator instance = new EmptyRangeIterator();
+        EmptyRangeIterator() { super(null, null, 0); }
+        public PrimaryKey computeNext() { return endOfData(); }
+        protected void performSkipTo(PrimaryKey nextToken) { }
+        public void close() { }
+    }
+
+    public static abstract class Builder
+    {
+        public enum IteratorType
+        {
+            CONCAT,
+            UNION,
+            INTERSECTION
+        }
+
+        @VisibleForTesting
+        protected final Statistics statistics;
+
+        @VisibleForTesting
+        protected final PriorityQueue<RangeIterator> ranges;
+
+        public Builder(IteratorType type)
+        {
+            statistics = new Statistics(type);
+            ranges = new PriorityQueue<>(16, Comparator.comparing(RangeIterator::getCurrent));
+        }
+
+        public PrimaryKey getMinimum()
+        {
+            return statistics.min;
+        }
+
+        public PrimaryKey getMaximum()
+        {
+            return statistics.max;
+        }
+
+        public long getTokenCount()
+        {
+            return statistics.tokenCount;
+        }
+
+        public int rangeCount()
+        {
+            return ranges.size();
+        }
+
+        public Collection<RangeIterator> ranges()
+        {
+            return ranges;
+        }
+
+        public Builder add(RangeIterator range)
+        {
+            if (range == null)
+                return this;
+
+            if (range.getCount() > 0)
+                ranges.add(range);
+            else
+                FileUtils.closeQuietly(range);
+            statistics.update(range);
+
+            return this;
+        }
+
+        public Builder add(List<RangeIterator> ranges)
+        {
+            if (ranges == null || ranges.isEmpty())
+                return this;
+
+            ranges.forEach(this::add);
+            return this;
+        }
+
+        public final RangeIterator build()
+        {
+            if (rangeCount() == 0)
+                return empty();
+            else
+                return buildIterator();
+        }
+
+        protected abstract RangeIterator buildIterator();
+
+        public static class Statistics
+        {
+            protected final IteratorType iteratorType;
+
+            protected PrimaryKey min, max;
+            protected long tokenCount;
```

Comment:
nit: Maybe just `count` or `keyCount` now. (I'd be fine w/ the former, given `min` and `max` don't prefix.) Also looks like a few instances of "token" in the JavaDoc below.